tests/test_main.py

```diff
@@ -0,0 +1,371 @@
+        )
+        assert crud_event_qs.count() == 1
+
+        obj.name = "changed name"
+        obj.save()
+        assert crud_event_qs.count() == 2
+
+        last_change = crud_event_qs.first()
+        assert "name" in last_change.changed_fields
+
+    def test_update(self, model):
+        obj = model.objects.create()
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        assert crud_event_qs.count() == 1
+
+        obj.name = "changed name"
+        obj.save()
+        assert crud_event_qs.count() == 2
+
+        last_change = crud_event_qs.first()
+        assert "name" in last_change.changed_fields
+
+    @pytest.mark.usefixtures("no_changed_fields_skip")
+    def test_fake_update_skip_no_changed_fields(self, model):
+        obj = model.objects.create()
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        obj.save()
+        assert crud_event_qs.count() == 1
+
+    def test_fake_update(self, model):
+        obj = model.objects.create()
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        obj.save()
+        assert crud_event_qs.count() == 2
+
+    def test_delete(self, model):
+        obj = model.objects.create()
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj.id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        assert crud_event_qs.count() == 1
+
+        obj_id = obj.pk
+        obj.delete()
+        crud_event_qs = CRUDEvent.objects.filter(
+            object_id=obj_id, content_type=ContentType.objects.get_for_model(obj)
+        )
+        assert crud_event_qs.count() == 2
+
+    @pytest.mark.usefixtures("_audit_logger")
+    def test_propagate_exceptions(self, model, settings):
+        with pytest.raises(ValueError, match="Test exception"):
+            model.objects.create()
+
+        settings.DJANGO_EASY_AUDIT_PROPAGATE_EXCEPTIONS = False
+        try:
+            model.objects.create()
+        except ValueError:
+            pytest.fail("Unexpected ValueError")
+
+
+class TestAuditUUIDModels(TestAuditModels):
+    @pytest.fixture
+    def model(self):
+        return UUIDModel
+
+    @pytest.fixture
+    def fk_model(self):
+        return UUIDForeignKeyModel
+
+    @pytest.fixture
+    def m2m_model(self):
+        return UUIDM2MModel
+
+
+class TestAuditBigIntModels(TestAuditModels):
+    @pytest.fixture
+    def model(self):
+        return BigIntModel
+
+    @pytest.fixture
+    def fk_model(self):
+        return BigIntForeignKeyModel
+
+    @pytest.fixture
+    def m2m_model(self):
+        return BigIntM2MModel
+
+
+@pytest.mark.django_db
+class TestMiddleware:
+    def test_middleware_logged_in(self, user, client, username, password):
```

Comment:
`pytest-django` provides a handy `client` fixture, so we no longer have to inherit from Django's `TestCase`: https://pytest-django.readthedocs.io/en/latest/helpers.html#client-django-test-client